Kernel/ACPI/DMIDecoder.cpp

```diff
@@ -0,0 +1,263 @@
+    mmap_region(*region, paddr);
+
+    SMBIOS::TableHeader* v_smbios_table = (SMBIOS::TableHeader*)(region->vaddr().get() + ((u32)physical_table & 0xfff));
+
+    u32 structures_count = 0;
+    while (table_length > 0) {
+#ifdef SMBIOS_DEBUG
+        kprintf("DMIDecoder: Examining table @ P 0x%x, V 0x%x\n", physical_table, v_smbios_table);
+#endif
+        structures_count++;
+        if (v_smbios_table->type == (u32)SMBIOS::TableType::EndOfTable) {
+            kprintf("DMIDecoder: Detected table with type 127, End of SMBIOS data.\n");
+            break;
+        } else {
+            m_smbios_tables.append(physical_table);
+            table_length -= v_smbios_table->length;
+            kprintf("DMIDecoder: Detected table with type %u\n", v_smbios_table->type);
+        }
+        SMBIOS::TableHeader* physical_next_table = get_next_physical_table(*physical_table);
+#ifdef SMBIOS_DEBUG
+        kprintf("DMIDecoder: Next table @ P 0x%x\n", physical_next_table);
+#endif
+        if (physical_next_table == nullptr) {
+            break;
+        } else {
+            v_smbios_table = (SMBIOS::TableHeader*)(region->vaddr().get() + (u32)physical_next_table - paddr.get());
+            physical_table = physical_next_table;
+        }
+    }
+    m_structures_count = structures_count;
+}
+
+size_t DMIDecoder::get_table_size(SMBIOS::TableHeader& table)
+{
+    // FIXME: Make sure we have some mapping here so we don't rely on existing identity mapping...
+    size_t i;
+#ifdef SMBIOS_DEBUG
+    kprintf("DMIDecoder: table legnth - 0x%x\n", table.length);
+#endif
+    const char* strtab = (char*)&table + table.length;
+    for (i = 1; strtab[i - 1] != '\0' || strtab[i] != '\0'; i++)
+        ;
+#ifdef SMBIOS_DEBUG
+    kprintf("DMIDecoder: table size - 0x%x\n", table.length + i + 1);
+#endif
+    return table.length + i + 1;
+}
+
+SMBIOS::TableHeader* DMIDecoder::get_next_physical_table(SMBIOS::TableHeader& p_table)
+{
+    return (SMBIOS::TableHeader*)((u32)&p_table + get_table_size(p_table));
+}
+
+SMBIOS::TableHeader* DMIDecoder::get_smbios_physical_table_by_handle(u16 handle)
+{
+    auto region = MM.allocate_kernel_region(PAGE_ROUND_UP(PAGE_SIZE * 2), "DMI Decoder Finding Table", Region::Access::Read);
+
+    for (auto* table : m_smbios_tables) {
+        if (table != nullptr) {
+            PhysicalAddress paddr = PhysicalAddress((u32)table & 0xffff000);
+            mmap_region(*region, paddr);
+            SMBIOS::TableHeader* table_v_ptr = (SMBIOS::TableHeader*)(region->vaddr().get() + ((u32)table & 0xfff));
+            if (table_v_ptr->handle == handle) {
+                return table;
+            }
+        }
+    }
+    return nullptr;
+}
+SMBIOS::TableHeader* DMIDecoder::get_smbios_physical_table_by_type(u8 table_type)
+{
+    auto region = MM.allocate_kernel_region(PAGE_ROUND_UP(PAGE_SIZE * 2), "DMI Decoder Finding Table", Region::Access::Read);
+
+    for (auto* table : m_smbios_tables) {
+        if (table != nullptr) {
```

Comment:
Please use early-fail to reduce nesting:

```suggestion
    if (!table)
        continue;
```

Reply:
That's actually a good idea :)